include/contractor/graph_contractor.hpp

```diff
@@ -133,6 +133,137 @@ class GraphContractor
         EnumerableThreadData data;
     };
 
+    EdgeWeight GetMinimalAdjacentWeight(const NodeID nid) const
+    {
+        const auto range = contractor_graph->GetAdjacentEdgeRange(nid);
+        if (range.empty())
+            return std::numeric_limits<EdgeWeight>::max() / 2;
+
+        const auto edge_weight = [this](const EdgeID edge_id) -> EdgeWeight {
+            return contractor_graph->GetEdgeData(edge_id).weight;
+        };
+
+        // cannot be std::min_elment, since edge_range does not fit the concept
+        auto minimum = edge_weight(range.front());
+        for (auto eid : range)
+            minimum = std::min(minimum, edge_weight(eid));
+        return minimum;
+    }
+
+    std::vector<EdgeWeight> BuildMinimalAdjacentWeightCache() const
+    {
+        const auto number_of_nodes = contractor_graph->GetNumberOfNodes();
+        std::vector<EdgeWeight> minimal_adjacent_weight_cache(number_of_nodes);
+
+        const decltype(number_of_nodes) grain_size = 10000;
+
+        tbb::parallel_for(
+            tbb::blocked_range<NodeID>(0, number_of_nodes, grain_size),
+            [this, &minimal_adjacent_weight_cache](const tbb::blocked_range<NodeID> &range) {
+                for (int x = range.begin(), end = range.end(); x != end; ++x)
+                    minimal_adjacent_weight_cache[x] = GetMinimalAdjacentWeight(x);
+            });
+
+        return minimal_adjacent_weight_cache;
+    }
+
+    /* Flush all data from the contraction to disc and reorder stuff for better locality */
+    void FlushDataAndRebuildContractorGraph(ThreadDataContainer &thread_data_list,
```

Comment:
Independently from whether this approach will work or not, I think this refactor makes sense.